docs/src/main/asciidoc/cdi-reference.adoc

```diff
@@ -168,6 +170,155 @@ public class CounterBean {
+* does not declare an observer,
+* does not declare any producer which is eligible for injection to any injection point,
+* is not directly eligible for injection into any `javax.enterprise.inject.Instance` or `javax.inject.Provider` injection point
+
+This optimization applies to all forms of bean declarations: bean class, producer method, producer field.
+
+Users can instruct the container to not remove any of their specific beans (even if they satisfy all the rules specified above) by annotating them with `io.quarkus.arc.Unremovable`.
+This annotation can be placed on the types, producer methods, and producer fields.
+ 
+Furthermore, extensions can eliminate possible false positives by producing `UnremovableBeanBuildItem`.
+
+Finally, Quarkus provides a middle ground for the bean removal optimization where application beans are never removed whether or not they are unused,
+while the optimization proceeds normally for non application classes. To use this mode, set `quarkus.arc.remove-unused-beans` to `fwk` or `framework`.
+
+When using the dev mode (running `mvn clean compile quarkus:dev`), you can see more information about which beans are being removed
+by enabling additional logging via the following line in your `application.properties`.
+
+----
+quarkus.log.category."io.quarkus.arc.processor".level=DEBUG
+----
+
+[[default_beans]]
+=== Default Beans
+
+Quarkus adds a capability that CDI currently does not support which is to conditionally declare a bean if no such bean has been declared in the typical ways Quarkus supports.
+This is done using the `@io.quarkus.arc.DefaultBean` annotation and is best explained with an example.
+
+Say there is a Quarkus extension that among other things declares a few CDI beans like the following code does:
+
+[source,java]
+----
+@Dependent
+public class TracerConfiguration {
+
+    @Produces
+    public Tracer tracer(Reporter reporter, Configuration configuration) {
+        return new Tracer(reporter, configuration);
+    }
+
+    @Produces
+    @DefaultBean
+    public Configuration configuration() {
+        // create a Configuration
+    }
+
+    @Produces
+    @DefaultBean
+    public Reporter reporter(){
+        // create a Reporter
+    }
+}
+----
+
+The idea is that the extension auto-configures things for user, eliminating a lot of boilerplate - we can just `@Inject` a `Tracer` wherever it is needed.
+Now imagine that in our application we would like to utilize the configured `Tracer`, but we need to customize it a little, for example by providing a custom `Reporter`.
+The only thing that would be needed in our application would be something like the following:
+
+
+[source,java]
+----
+@Dependent
+public class CustomTracerConfiguration {
+
+    @Produces
+    public Reporter reporter(){
+        // create a custom Reporter
+    }
+}
+----
+
+`@DefaultBean` allows extensions (or any other code for that matter) to provide defaults while backing off if beans of that type are supplied in any
```

Comment:
This is a repetition of the above I think it could removed.

Reply:
@mkouba apparently, I cannot suggest a multi line code change, but i would still delete this sentence as I think it just repeats what is said/implied from text and code above.